chain/src/chain.rs

```diff
@@ -117,7 +126,11 @@ impl<CI: ChainIndex + 'static> ChainService<CI> {
         }
     }
 
-    fn process_block(&mut self, block: Arc<Block>) -> Result<(), ProcessBlockError> {
+    fn process_block(
+        &mut self,
+        block: Arc<Block>,
+        cycles_set: Vec<Cycle>,
```

Comment:
This `cycles_set` parameter is confused me, it spent my lots of time to understand the behaviour:
if cycles_set[index] equals to Cycle::default() then it will use max_cycles to verify then rewrite the cycles to cycles_set, otherwise it use cycles_set[index] as max_cycles to execute validation.

It is strange, do we really need such behavior?
A better and simple solution is computing the actual cycles in chain and pool and return the cycles to let the invoker to decide whether it is valid or not.

Reply:
We only need to compare the actual txs cycles to the txs cycles we received from network peers to determine a peer is honest or lying. The Pool and Chain should cache the cycles, but it is not necessary to pass a cycles limit as a parameter.

Reply:
The cycles_set records the cycles actually used of all transactions, if it is Cycle::default(), it means that the transaction has not been executed and the value is unknown. If it is not Cycle::default(), the verification will skiped. We need a cycles_set because if we want to readd the transaction in the pool when switch fork, we need know the cycles.

Reply:
The verification behavior is different.